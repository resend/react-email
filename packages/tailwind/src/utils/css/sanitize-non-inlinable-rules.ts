import { type CssNode, string, walk } from 'css-tree';
import { sanitizeClassName } from '../compatibility/sanitize-class-name';
import { isRuleInlinable } from './is-rule-inlinable';

/**
 * This function goes through a few steps to ensure the best email client support and
 * to ensure that the media queries and pseudo classes are going to applied correctly alongside
 * the inline styles.
 *
 * What it does is:
 * 1. Converts all declarations in all rules into being important ones
 * 2. Sanitizes all the selectors of all non-inlinable rules
 */
export function sanitizeNonInlinableRules(node: CssNode) {
  walk(node, {
    visit: 'Rule',
    enter(rule) {
      if (!isRuleInlinable(rule)) {
        walk(rule.prelude, (node) => {
          if (node.type === 'ClassSelector') {
            const unescapedClassName = string.decode(node.name);
            node.name = sanitizeClassName(unescapedClassName);
          }
        });

        walk(rule, {
          visit: 'Declaration',
          enter(declaration) {
            declaration.important = true;
          },
        });
      }
    },
  });
};
