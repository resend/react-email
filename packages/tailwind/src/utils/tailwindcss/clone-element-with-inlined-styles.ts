import type { Node } from 'postcss';
import React from 'react';
import type { EmailElementProps } from '../../tailwind';
import { sanitizeClassName } from '../compatibility/sanitize-class-name';
import { makeInlineStylesFor } from '../css/make-inline-styles-for';
import { sanitizeMediaQueries } from '../css/media-queries/sanitize-media-queries';
import { sanitizeDeclarations } from '../css/sanitize-declarations';
import { isComponent } from '../react/is-component';
import type { setupTailwind } from './setup-tailwind';

export const cloneElementWithInlinedStyles = (
  element: React.ReactElement<EmailElementProps>,
  tailwind: ReturnType<typeof setupTailwind>,
) => {
  const propsToOverwrite: Partial<EmailElementProps> = {};

  let nonInlinableClasses: string[] = [];
  let nonInlineStyleNodes: Node[] = [];

  if (element.props.className) {
    const rootForClasses = tailwind.generateRootForClasses(
      element.props.className.split(' '),
    );
    sanitizeDeclarations(rootForClasses);

    const { sanitizedAtRules, mediaQueryClasses } =
      sanitizeMediaQueries(rootForClasses);
    nonInlinableClasses = mediaQueryClasses;
    nonInlineStyleNodes = sanitizedAtRules;

    const { styles, residualClassName } = makeInlineStylesFor(
      element.props.className,
      rootForClasses,
    );
    propsToOverwrite.style = {
      ...styles,
      ...element.props.style,
    };

    if (!isComponent(element)) {
      if (residualClassName.trim().length > 0) {
        propsToOverwrite.className = residualClassName;

        /*
          We sanitize only the class names of Tailwind classes that we are not going to inline
          to avoid unpredictable behavior on the user's code. If we did sanitize all classes
          a user-defined class could end up also being sanitized which would lead to unexpected 
          behavior and bugs that are hard to track.
        */
        for (const singleClass of mediaQueryClasses) {
          propsToOverwrite.className = propsToOverwrite.className.replace(
            singleClass,
            sanitizeClassName(singleClass),
          );
        }
      } else {
        propsToOverwrite.className = undefined;
      }
    }
  }

  const newProps = {
    ...element.props,
    ...propsToOverwrite,
  };

  return {
    elementWithInlinedStyles: React.cloneElement(
      element,
      newProps,
      newProps.children,
    ),

    nonInlinableClasses,
    nonInlineStyleNodes,
  };
};
