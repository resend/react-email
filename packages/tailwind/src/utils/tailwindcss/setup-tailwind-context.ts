import { createContext } from 'tailwindcss/lib/lib/setupContextUtils';
import resolveConfig from 'tailwindcss/resolveConfig';
import type { TailwindConfig } from '../../tailwind';

export const setupTailwindContext = (config: TailwindConfig) => {
  return createContext(
    resolveConfig({
      ...config,
      content: [],
      corePlugins: {
        preflight: false,
      },
    }),
  );
};
