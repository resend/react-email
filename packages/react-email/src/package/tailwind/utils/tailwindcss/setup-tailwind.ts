import type { Rule } from 'patched-postcss';
import parse from 'patched-postcss/lib/parse';
import collapseAdjacentRules from 'patched-tailwindcss/lib/lib/collapseAdjacentRules';
import collapseDuplicateDeclarations from 'patched-tailwindcss/lib/lib/collapseDuplicateDeclarations';
import evaluateTailwindFunctions from 'patched-tailwindcss/lib/lib/evaluateTailwindFunctions';
import expandApplyAtRules from 'patched-tailwindcss/lib/lib/expandApplyAtRules';
import expandTailwindAtRules from 'patched-tailwindcss/lib/lib/expandTailwindAtRules';
import { generateRules as rawGenerateRules } from 'patched-tailwindcss/lib/lib/generateRules';
import partitionApplyAtRules from 'patched-tailwindcss/lib/lib/partitionApplyAtRules';
import resolveDefaultsAtRules from 'patched-tailwindcss/lib/lib/resolveDefaultsAtRules';
import substituteScreenAtRules from 'patched-tailwindcss/lib/lib/substituteScreenAtRules';
import type { TailwindConfig } from '../../tailwind';
import { resolveAllCSSVariables } from '../css/resolve-all-css-variables';
import { setupTailwindContext } from './setup-tailwind-context';

const tailwindAtRulesRoot = parse(
  `
  @tailwind base;
  @tailwind components;
`,
).root();

export function setupTailwind(config: TailwindConfig) {
  const tailwindContext = setupTailwindContext(config);
  return {
    generateRootForClasses: (classes: string[]) => {
      const bigIntRuleTuples: [bigint, Rule][] = rawGenerateRules(
        new Set(classes),
        tailwindContext,
      );

      const root = tailwindAtRulesRoot
        .clone()
        .append(...bigIntRuleTuples.map(([, rule]) => rule));
      partitionApplyAtRules()(root);
      // This is fine because the internal await is never actually called out
      // because of there not being any `changedContent` files on the context
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      expandTailwindAtRules(tailwindContext)(root);
      partitionApplyAtRules()(root);
      expandApplyAtRules(tailwindContext)(root);
      evaluateTailwindFunctions(tailwindContext)(root);
      substituteScreenAtRules(tailwindContext)(root);
      resolveDefaultsAtRules(tailwindContext)(root);
      collapseAdjacentRules(tailwindContext)(root);
      collapseDuplicateDeclarations(tailwindContext)(root);

      resolveAllCSSVariables(root);

      return root;
    },
  };
}
