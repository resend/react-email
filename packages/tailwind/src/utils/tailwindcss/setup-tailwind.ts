import type { Rule } from 'postcss';
import { parse } from 'postcss';
import collapseAdjacentRules from 'tailwindcss/lib/lib/collapseAdjacentRules';
import collapseDuplicateDeclarations from 'tailwindcss/lib/lib/collapseDuplicateDeclarations';
import evaluateTailwindFunctions from 'tailwindcss/lib/lib/evaluateTailwindFunctions';
import expandApplyAtRules from 'tailwindcss/lib/lib/expandApplyAtRules';
import expandTailwindAtRules from 'tailwindcss/lib/lib/expandTailwindAtRules';
import { generateRules as rawGenerateRules } from 'tailwindcss/lib/lib/generateRules';
import partitionApplyAtRules from 'tailwindcss/lib/lib/partitionApplyAtRules';
import resolveDefaultsAtRules from 'tailwindcss/lib/lib/resolveDefaultsAtRules';
import substituteScreenAtRules from 'tailwindcss/lib/lib/substituteScreenAtRules';
import type { TailwindConfig } from '../../tailwind';
import { resolveAllCSSVariables } from '../css/resolve-all-css-variables';
import { setupTailwindContext } from './setup-tailwind-context';

const tailwindAtRulesRoot = parse(
  `
  @tailwind base;
  @tailwind components;
`,
).root();

export function setupTailwind(config: TailwindConfig) {
  const tailwindContext = setupTailwindContext(config);
  return {
    generateRootForClasses: (classes: string[]) => {
      const bigIntRuleTuples: [bigint, Rule][] = rawGenerateRules(
        new Set(classes),
        tailwindContext,
      );

      const root = tailwindAtRulesRoot
        .clone()
        .append(...bigIntRuleTuples.map(([, rule]) => rule));
      partitionApplyAtRules()(root);
      // This is fine because the internal await is never actually called out
      // because of there not being any `changedContent` files on the context
      // eslint-disable-next-line @typescript-eslint/no-floating-promises
      expandTailwindAtRules(tailwindContext)(root);
      partitionApplyAtRules()(root);
      expandApplyAtRules(tailwindContext)(root);
      evaluateTailwindFunctions(tailwindContext)(root);
      substituteScreenAtRules(tailwindContext)(root);
      resolveDefaultsAtRules(tailwindContext)(root);
      collapseAdjacentRules(tailwindContext)(root);
      collapseDuplicateDeclarations(tailwindContext)(root);

      resolveAllCSSVariables(root);

      return root;
    },
  };
}
