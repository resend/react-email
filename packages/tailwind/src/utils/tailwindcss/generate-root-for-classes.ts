import { parse } from 'css-tree';
import { type Config, compile } from 'tailwindcss';
import { resolveAllCSSVariables } from '../css/resolve-all-css-variables';
import indexCss from './tailwind-stylesheets/index';
import preflightCss from './tailwind-stylesheets/preflight';
import themeCss from './tailwind-stylesheets/theme';
import utilitiesCss from './tailwind-stylesheets/utilities';

const baseCss = `
@layer theme, base, components, utilities;
@import "tailwindcss/theme.css" layer(theme);
@import "tailwindcss/utilities.css" layer(utilities);
@config;
`;

export async function generateRootForClasses(
  classes: string[],
  config: Config,
) {
  const compiler = await compile(baseCss, {
    async loadModule(id, base, resourceHint) {
      if (resourceHint === 'config') {
        return {
          path: id,
          base: base,
          module: config,
        };
      }

      throw new Error(
        `NO-OP: should we implement support for ${resourceHint}?`,
      );
    },
    async loadStylesheet(id, base) {
      if (id === 'tailwindcss') {
        return {
          base,
          path: 'tailwindcss/index.css',
          content: indexCss,
        };
      }

      if (id === 'tailwindcss/preflight.css') {
        return {
          base,
          path: id,
          content: preflightCss,
        };
      }

      if (id === 'tailwindcss/theme.css') {
        return {
          base,
          path: id,
          content: themeCss,
        };
      }

      if (id === 'tailwindcss/utilities.css') {
        return {
          base,
          path: id,
          content: utilitiesCss,
        };
      }

      throw new Error(
        'stylesheet not supported, you can only import tailwindcss',
      );
    },
  });
  const css = compiler.build(classes);
  const root = parse(css);
  resolveAllCSSVariables(root);

  return root;
}
