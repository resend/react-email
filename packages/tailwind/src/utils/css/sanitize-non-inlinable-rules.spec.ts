import { generate } from 'css-tree';
import { setupTailwind } from '../tailwindcss/setup-tailwind';
import { sanitizeNonInlinableRules } from './sanitize-non-inlinable-rules';

describe('sanitizeNonInlinableRules()', () => {
  it('should handle rules that can be inlined', async () => {
    const tailwind = await setupTailwind({});
    const stylesheet = tailwind.addUtilities(
      ['bg-gray-900', 'text-red-300', 'text-lg'],
      { compatibilityFixes: false },
    );

    sanitizeNonInlinableRules(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();
  });

  it('should css nesting in hover pseudo styles', async () => {
    const tailwind = await setupTailwind({});
    const stylesheet = tailwind.addUtilities(
      [
        'hover:text-sky-600',
        'sm:focus:outline-none',
        'md:hover:bg-gray-100',
        'lg:focus:underline',
      ],
      { compatibilityFixes: false },
    );

    sanitizeNonInlinableRules(stylesheet);
    expect(generate(stylesheet)).toMatchSnapshot();
  });

  it('shuold work with basic media query rules', async () => {
    const tailwind = await setupTailwind({});
    const stylesheet = tailwind.addUtilities(
      ['sm:mx-auto', 'sm:max-w-lg', 'sm:rounded-lg', 'md:px-10', 'md:py-12'],
      { compatibilityFixes: false },
    );

    sanitizeNonInlinableRules(stylesheet);

    expect(generate(stylesheet)).toMatchSnapshot();
  });
});
