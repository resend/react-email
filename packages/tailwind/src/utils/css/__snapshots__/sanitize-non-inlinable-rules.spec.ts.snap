// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sanitizeNonInlinableRules() > should css nesting in hover pseudo styles 1`] = `"/*! tailwindcss v4.1.12 | MIT License | https://tailwindcss.com */@layer theme,base,components,utilities;@layer theme{:root,:host{--color-sky-600: oklch(58.8% 0.158 241.966)!important;--color-gray-100: oklch(96.7% 0.003 264.542)!important}}@layer utilities{.hover_text-sky-600{&:hover{@media (hover:hover){color:var(--color-sky-600)!important}}}.sm_focus_outline-none{@media (width>=40rem){&:focus{--tw-outline-style: none!important;outline-style:none!important}}}.md_hover_bg-gray-100{@media (width>=48rem){&:hover{@media (hover:hover){background-color:var(--color-gray-100)!important}}}}.lg_focus_underline{@media (width>=64rem){&:focus{text-decoration-line:underline!important}}}}"`;

exports[`sanitizeNonInlinableRules() > should handle rules that can be inlined 1`] = `"/*! tailwindcss v4.1.12 | MIT License | https://tailwindcss.com */@layer theme,base,components,utilities;@layer theme{:root,:host{--color-red-300: oklch(80.8% 0.114 19.571)!important;--color-gray-900: oklch(21% 0.034 264.665)!important;--text-lg: 1.125rem!important;--text-lg--line-height: calc(1.75 / 1.125)!important}}@layer utilities{.bg-gray-900{background-color:var(--color-gray-900)}.text-lg{font-size:var(--text-lg);line-height:var(--tw-leading, var(--text-lg--line-height))}.text-red-300{color:var(--color-red-300)}}"`;

exports[`sanitizeNonInlinableRules() > should work with basic media query rules 1`] = `"/*! tailwindcss v4.1.12 | MIT License | https://tailwindcss.com */@layer theme,base,components,utilities;@layer theme{:root,:host{--spacing: 0.25rem!important;--container-lg: 32rem!important;--radius-lg: 0.5rem!important}}@layer utilities{.sm_mx-auto{@media (width>=40rem){margin-inline:auto!important}}.sm_max-w-lg{@media (width>=40rem){max-width:var(--container-lg)!important}}.sm_rounded-lg{@media (width>=40rem){border-radius:var(--radius-lg)!important}}.md_px-10{@media (width>=48rem){padding-inline:calc(var(--spacing)*10)!important}}.md_py-12{@media (width>=48rem){padding-block:calc(var(--spacing)*12)!important}}}"`;
